File: ports/jagentic-core/src/test/java/org/jagentic/core/ContextWindowTest.java
package org.jagentic.core;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertSame;
import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.UUID;
import java.util.concurrent.ThreadLocalRandom;

import org.junit.jupiter.api.Test;

/**
 * {@code context.compaction: window}: the fold retains the most recent {@code max_items} messages in
 * log order, {@code transcript_length} reports what is retained, the log and {@code turn_count} are
 * untouched, and {@code none} retains everything.
 */
class ContextWindowTest {

  private static String rnd() {
    return UUID.randomUUID().toString().substring(0, 8);
  }

  /** One completed turn: turn_received, memory_written (two messages), turn_completed. */
  private static void completeTurn(ConversationLog log, String cid, int i) {
    String tid = "t" + i;
    log.append(cid, tid, EventType.TURN_RECEIVED, Map.of("turn_id", tid, "text", "u" + i));
    List<Map<String, Object>> messages = new ArrayList<>();
    messages.add(Map.of("role", "user", "text", "u" + i));
    messages.add(Map.of("role", "assistant", "text", "a" + i));
    log.append(cid, tid, EventType.MEMORY_WRITTEN, Map.of("messages", messages));
    log.append(cid, tid, EventType.TURN_COMPLETED, Map.of("reply", "a" + i));
  }

  @Test
  void windowRetainsTheMostRecentMaxItemsInOrderAndCountsAllTurns() {
    ThreadLocalRandom r = ThreadLocalRandom.current();
    int maxItems = r.nextInt(1, 9);
    int turns = r.nextInt(1, 12);
    int written = 2 * turns;
    ConversationLog log = new ConversationLog.InMemory();
    String cid = "c-" + rnd();
    for (int i = 0; i < turns; i++) {
      completeTurn(log, cid, i);
    }
    ConversationState full = log.state(cid);
    ConversationState windowed = log.state(cid, ContextWindow.window(maxItems));

    assertEquals(written, full.transcriptLength(), "no window keeps every message");
    assertEquals(Math.min(written, maxItems), windowed.transcriptLength());
    assertEquals(Math.min(written, maxItems), windowed.transcript().size());
    assertEquals(turns, windowed.turnCount(), "turn_count is unaffected by compaction");
    assertEquals(full.turnCount(), windowed.turnCount());
    assertEquals(full.transcript().subList(written - windowed.transcript().size(), written),
        windowed.transcript(), "the retained messages are the most recent, in log order");
    assertEquals(3L * turns, log.events(cid).size(), "the log itself is never compacted");
    assertEquals(full.turns(), windowed.turns(), "per-turn records are unaffected");
    Map<String, Object> expected = new LinkedHashMap<>();
    expected.put("turn_count", (long) turns);
    expected.put("transcript_length", (long) Math.min(written, maxItems));
    assertEquals(expected, windowed.reduced());
  }

  @Test
  void conversationsAreWindowedIndependently() {
    ThreadLocalRandom r = ThreadLocalRandom.current();
    int maxItems = r.nextInt(2, 7);
    int turnsA = r.nextInt(maxItems, maxItems + 6);
    int turnsB = r.nextInt(1, maxItems);
    ConversationLog log = new ConversationLog.InMemory();
    String a = "a-" + rnd();
    String b = "b-" + rnd();
    for (int i = 0; i < turnsA; i++) {
      completeTurn(log, a, i);
    }
    for (int i = 0; i < turnsB; i++) {
      completeTurn(log, b, i);
    }
    ContextWindow w = ContextWindow.window(maxItems);
    assertEquals(maxItems, log.state(a, w).transcriptLength());
    assertEquals(turnsA, log.state(a, w).turnCount());
    assertEquals(Math.min(2 * turnsB, maxItems), log.state(b, w).transcriptLength());
    assertEquals(turnsB, log.state(b, w).turnCount());
  }

  @Test
  void noneAndMoscowLeaveTheRetainedTranscriptAlone() {
    int turns = ThreadLocalRandom.current().nextInt(1, 10);
    ConversationLog log = new ConversationLog.InMemory();
    String cid = "c-" + rnd();
    for (int i = 0; i < turns; i++) {
      completeTurn(log, cid, i);
    }
    ConversationState none = log.state(cid, ContextWindow.fromMap(Map.of("compaction", "none", "max_items", 1)));
    ConversationState moscow = log.state(cid, ContextWindow.fromMap(Map.of("compaction", "moscow", "max_tokens", 8)));
    assertEquals(2L * turns, none.transcriptLength());
    assertEquals(2L * turns, moscow.transcriptLength());
    assertEquals(log.state(cid), none);
    assertEquals(log.state(cid), moscow);
    assertSame(ContextWindow.NONE, ContextWindow.fromMap(null));
    assertFalse(ContextWindow.fromMap(Map.of("max_items", 3)).bounded(), "compaction defaults to none");
  }

  @Test
  void retainKeepsTheTailAndFromMapValidatesTheWindow() {
    ThreadLocalRandom r = ThreadLocalRandom.current();
    int maxItems = r.nextInt(1, 6);
    int n = r.nextInt(maxItems, maxItems + 10);
    List<Integer> items = new ArrayList<>();
    for (int i = 0; i < n; i++) {
      items.add(i);
    }
    ContextWindow w = ContextWindow.fromMap(Map.of("compaction", "window", "max_items", maxItems));
    assertTrue(w.bounded());
    assertEquals(items.subList(n - maxItems, n), w.retain(items));
    assertEquals(maxItems, w.retainedLength(n));
    assertEquals(0, w.retainedLength(0));
    assertSame(items, ContextWindow.NONE.retain(items));
    assertThrows(IllegalArgumentException.class, () -> ContextWindow.fromMap(Map.of("compaction", "window")));
    assertThrows(IllegalArgumentException.class, () -> ContextWindow.window(0));
  }

  @Test
  void aRunningWorkflowReportsTheRetainedWindowInResultState() {
    ThreadLocalRandom r = ThreadLocalRandom.current();
    int maxItems = r.nextInt(1, 7);
    int turns = r.nextInt(1, 8);
    Map<String, Object> path = new LinkedHashMap<>();
    path.put("brain", "rule");
    path.put("prompt", "You chat.");
    Map<String, Object> agent = new LinkedHashMap<>();
    agent.put("id", "chat");
    agent.put("router", Map.of("kind", "keyword", "default", "main"));
    agent.put("paths", Map.of("main", path));
    Map<String, Object> spec = new LinkedHashMap<>();
    spec.put("spec_version", "agentic/v1");
    spec.put("backend", "local");
    spec.put("context", Map.of("max_items", maxItems, "compaction", "window"));
    spec.put("agent", agent);
    org.jagentic.core.pipeline.GraphBuilder.Built built =
        org.jagentic.core.pipeline.GraphBuilder.build(spec, null);
    assertEquals(ContextWindow.window(maxItems), built.graph().contextWindow());
    ConversationLog log = new ConversationLog.InMemory();
    LocalRuntime rt = new LocalRuntime(built.graph(), new ConversationStore.InMemory(),
        new KeyedStateStore.InMemory(), built.tools(), built.retriever(), log);
    String cid = "c-" + rnd();
    TurnResult last = null;
    for (int i = 1; i <= turns; i++) {
      last = rt.submit(Event.turn(cid, "t" + i, "u", "hello " + rnd()));
      assertEquals(TurnStatus.COMPLETED, last.status);
      assertEquals((long) i, last.state.get("turn_count"));
      assertEquals((long) Math.min(2 * i, maxItems), last.state.get("transcript_length"));
    }
    assertEquals(2L * turns, log.state(cid).transcriptLength(), "the log keeps every message");
    assertEquals(last.state, rt.replay(cid).reduced(), "replay reads the retained window too");
  }
}

File: agentic-clj/test/agentic/context_window_test.clj
(ns agentic.context-window-test
  "`context.compaction: window`: the fold retains the most recent `max_items` messages in log order,
   `transcript_length` reports what is retained, the log and `turn_count` are untouched, `none`
   retains everything, and conversations are windowed independently."
  (:require [clojure.test :refer [deftest is testing]]
            [agentic.pipeline :as pipeline]
            [agentic.core :as core]
            [agentic.log :as log]))

(defn- rand-between [lo hi] (+ lo (rand-int (inc (- hi lo)))))

(defn- completed-turn
  "The events of one completed turn: turn_received, memory_written (two messages), turn_completed."
  [i]
  [{:turn-id (str "t" i) :type :turn-received :payload {:turn-id (str "t" i) :text (str "u" i)}}
   {:turn-id (str "t" i) :type :memory-written
    :payload {:messages [{:role "user" :text (str "u" i)} {:role "assistant" :text (str "a" i)}]}}
   {:turn-id (str "t" i) :type :turn-completed :payload {:reply (str "a" i)}}])

(defn- log-of [turns]
  (vec (map-indexed (fn [i e] (assoc e :sequence i)) (mapcat completed-turn (range turns)))))

(deftest window-bounds-transcript-length-not-turn-count
  (dotimes [_ 20]
    (let [max-items (rand-between 1 8)
          turns (rand-between 1 11)
          written (* 2 turns)
          events (log-of turns)
          window {:max-items max-items :compaction "window"}
          full (log/reduce-state events)
          bounded (log/reduce-state events window)]
      (is (= written (:transcript-length full)) "no context keeps every message")
      (is (= (min written max-items) (:transcript-length bounded)))
      (is (= turns (:turn-count full) (:turn-count bounded)) "turn_count is unaffected by compaction")
      (is (= (* 3 turns) (count events)) "the log itself is never compacted")
      (is (= full (log/reduce-state events {:max-items 1 :compaction "none"})) "none retains everything")
      (is (= full (log/reduce-state events {:max-tokens 8 :compaction "moscow"})) "moscow is not a message window"))))

(deftest retain-window-keeps-the-most-recent-messages-in-order
  (dotimes [_ 20]
    (let [max-items (rand-between 1 6)
          n (rand-between max-items (+ max-items 10))
          messages (mapv (fn [i] {:role (if (even? i) "user" "assistant") :content (str "m" i)}) (range n))
          kept (log/retain-window messages {:max-items max-items :compaction "window"})]
      (is (= (subvec messages (- n max-items)) kept))
      (is (= messages (log/retain-window messages nil)))
      (is (= messages (log/retain-window messages {:max-items 1 :compaction "none"})))))
  (is (thrown? clojure.lang.ExceptionInfo (log/window-size {:compaction "window"})))
  (is (thrown? clojure.lang.ExceptionInfo (log/window-size {:compaction "window" :max-items 0})))
  (is (nil? (log/window-size nil)))
  (is (nil? (log/window-size {:max-items 3}))))

(defn- windowed-workflow [max-items]
  {"spec_version" "agentic/v1" "backend" "local"
   "context" {"max_items" max-items "compaction" "window"}
   "agent" {"id" "chat"
            "router" {"kind" "keyword" "default" "main"}
            "paths" {"main" {"brain" "rule" "prompt" "You chat."}}}})

(deftest running-workflow-reports-the-retained-window
  (let [max-items (rand-between 1 6)
        turns (rand-between 1 7)
        sys (pipeline/system-from (windowed-workflow max-items) {})
        results (mapv #(core/submit sys {:conversation-id "c1" :turn-id (str "t" %) :user-id "u" :text (str "hello " %)})
                      (range 1 (inc turns)))
        other (core/submit sys {:conversation-id "c2" :turn-id "x1" :user-id "u" :text "other conversation"})]
    (testing "every turn completes and reports the retained transcript"
      (is (every? #(= :completed (:status %)) results))
      (is (= (range 1 (inc turns)) (map #(get-in % [:state :turn-count]) results)))
      (is (= (map #(min (* 2 %) max-items) (range 1 (inc turns)))
             (map #(get-in % [:state :transcript-length]) results))))
    (testing "the log keeps every message and the folded state reads the window"
      (is (= (* 2 turns) (:transcript-length (log/reduce-state (core/events sys "c1")))))
      (is (= {:turn-count turns :transcript-length (min (* 2 turns) max-items)} (core/state sys "c1"))))
    (testing "conversations are windowed independently"
      (is (= {:turn-count 1 :transcript-length (min 2 max-items)} (:state other))))))
